cmd/metal-api/internal/grpc/boot-service.go

```diff
@@ -0,0 +1,411 @@
+package grpc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/dustin/go-humanize"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/datastore"
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	v1 "github.com/metal-stack/metal-api/pkg/api/v1"
+	"github.com/metal-stack/metal-lib/bus"
+	"go.uber.org/zap"
+)
+
+type BootService struct {
+	log          *zap.SugaredLogger
+	ds           Datasource
+	publisher    bus.Publisher
+	eventService *EventService
+}
+
+func NewBootService(cfg *ServerConfig, eventService *EventService) *BootService {
+	return &BootService{
+		ds:           cfg.Datasource,
+		log:          cfg.Logger.Named("boot-service"),
+		publisher:    cfg.Publisher,
+		eventService: eventService,
+	}
+}
+func (b *BootService) Dhcp(ctx context.Context, req *v1.BootServiceDhcpRequest) (*v1.BootServiceDhcpResponse, error) {
+	b.log.Infow("dhcp", "req", req)
+
+	_, err := b.eventService.Send(ctx, &v1.EventServiceSendRequest{Events: map[string]*v1.MachineProvisioningEvent{
+		req.Uuid: {
+			Event:   string(metal.ProvisioningEventPXEBooting),
+			Message: "machine sent extended dhcp request",
+		},
+	}})
+	if err != nil {
+		return nil, err
+	}
+	return &v1.BootServiceDhcpResponse{}, nil
+}
+
+func (b *BootService) Boot(ctx context.Context, req *v1.BootServiceBootRequest) (*v1.BootServiceBootResponse, error) {
+	b.log.Infow("boot", "req", req)
+	if req == nil {
+		return nil, fmt.Errorf("req is nil")
+	}
+
+	var m *metal.Machine
+	err := b.ds.FindMachine(&datastore.MachineSearchQuery{
+		NicsMacAddresses: []string{req.Mac},
+		PartitionID:      &req.PartitionId,
+	}, m)
+	if err != nil && !metal.IsNotFound(err) {
+		return nil, err
+	}
+
+	if m != nil && m.PartitionID != req.PartitionId {
+		return nil, fmt.Errorf("partitionID:%q of machine with mac does not match partitionID:%q", m.PartitionID, req.PartitionId)
+	}
+	p, err := b.ds.FindPartition(req.PartitionId)
+	if err != nil || p == nil {
+		return nil, fmt.Errorf("no partition with id:%q found %w", req.PartitionId, err)
+	}
+
+	resp := &v1.BootServiceBootResponse{
+		Kernel:       p.BootConfiguration.KernelURL,
+		InitRamDisks: []string{p.BootConfiguration.ImageURL},
+		Cmdline:      &p.BootConfiguration.CommandLine,
+	}
+	b.log.Infow("boot", "resp", resp)
+	return resp, nil
+
+	// if allocateion.Succeed==false, the machine was already in the installation phase but crashed before finalizing allocation
```

Comment:
Needs to be discussed, code was taken from metal-core impl.

Reply:
I believe that we wanted to be very safe that a machine when the metal-hammer gets booted and the machine still belongs to a user that we do not wipe the disks. But somehow it does not make sense and it also never worked (see #209). I think we can remove it.

Reply:
removed commented code